### PR TITLE
DPL: refactoring of WorkflowSerializationHelpers

### PR DESCRIPTION
**Pull request by matthiasrichter**
Preparing for better serialization of DataMatchers
- new states added to importer
- json import of InputSpec now based on stack of DataMatchers

### CODE REVIEW

File: Framework/Core/src/WorkflowSerializationHelpers.cxx [hunk 1]
```diff
@@ -277,34 +298,28 @@ struct WorkflowImporter : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>,
...
-          std::make_unique<DataDescriptorMatcher>(
-            DataDescriptorMatcher::Op::And,
-            DescriptionValueMatcher{ContextRef{1}},
-            std::make_unique<DataDescriptorMatcher>(
-              DataDescriptorMatcher::Op::And,
-              SubSpecificationTypeValueMatcher{ContextRef{2}},
-              std::make_unique<DataDescriptorMatcher>(DataDescriptorMatcher::Op::Just,
-                                                      StartTimeValueMatcher{ContextRef{0}})))};
-
-        dataProcessors.back().inputs.push_back(InputSpec({binding}, std::move(expectedMatcher00)));
```
Comment (matthiasrichter) [inline Framework/Core/src/WorkflowSerializationHelpers.cxx:294]: @ktf in this case, the `lifetime` and `inputOptions` parameters were missing, was this a bug? The two other cases below are using those two additional parameters.

File: Framework/Core/src/WorkflowSerializationHelpers.cxx [hunk 2]
```diff
@@ -277,34 +298,28 @@ struct WorkflowImporter : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>,
...
-      } else if (inputHasDescription) {
-        if (inputHasSubSpec) {
-          dataProcessors.back().inputs.push_back(InputSpec(binding, origin, description, subspec, lifetime, inputOptions));
-        } else {
-          dataProcessors.back().inputs.push_back(InputSpec(binding, {origin, description}, lifetime, inputOptions));
+      auto buildMatcher = [](auto& nodes) -> std::unique_ptr<DataDescriptorMatcher> {
+        auto lastMatcher =
+          std::make_unique<DataDescriptorMatcher>(DataDescriptorMatcher::Op::Just,
+                                                  StartTimeValueMatcher(ContextRef{0}));
```
Comment (matthiasrichter) [inline Framework/Core/src/WorkflowSerializationHelpers.cxx]: `ContextRef{0}` is always reserved for start time value matching, right?

Comment (ktf) [reply Framework/Core/src/WorkflowSerializationHelpers.cxx]: Yes. We should probably change it to `data_matcher::STARTTIME_POS`.

Comment (matthiasrichter) [reply Framework/Core/src/WorkflowSerializationHelpers.cxx]: ok, will do

**Comment by ktf** [comment]
Issue looks genuine?

**Comment by matthiasrichter** [comment]
Need to look into the error of the QC test
```
[ERROR] invalid workflow in o2-qc: InputSpec for Enumeration must be fully qualified
```

**Comment by matthiasrichter** [comment]
The issue comes from the fact, that some input specs are required to be fully qualified, and this is implemented by requiring the input spec to be of type `ConcreteDataMatcher`. This is not the case with this PR, this only places `DataDescriptorMatcher` objects in the input spec when restoring the workflow from json configuration. The matchers still describe fully qualified specs, but in a different representation. The logic needs to be revised.

Since this is becoming very complex, as a simple solution a helper method is needed which creates a `ConcreteDataMatcher` from a `DataDescriptorMatcher` object if this is describing a fully qualified data spec with unique origin/description/subspec and only `and` operations. The available helpers, e.g. `extractMatcherInfo` are not handling more complex matcher objects. To move on with this task, the first implementation will only support a fixed DataDescriptorMatcher hierarchy, and later we can extend also the other helpers.

**Comment by matthiasrichter** [comment]
with merge #8677, the original behavior of adding simplified ConcreteDataMatcher objects has been restored in this PR.

**Comment by matthiasrichter** [comment]
The remaining CI issue is unrelated,
```
[ERROR] Requested resource does not exist: http://ccdb-test.cern.ch:8080/Test/CachingA/500/
```
makes `CCDB/test/testBasicCCDBManager.cxx` failing

**Pull request merged**